server/audit_math/suite.py

```diff
@@ -1,718 +0,0 @@
-        contest: the contest being audited
-        bp_stratum: a stratum object containing the ballot polling stratum
-        cvr_stratum: a stratum object containing the ballot comparison stratum
-        winner, loser: the winner-loser pair to maximuze the p-value over
-
-    Outputs:
-        a maximum p-value for the Fisher's combining function given the two
-        strata.
-    """
-    maximized_pvalue = 0.0
-    # find range of possible lambda
-    cvr_winner_votes = cvr_stratum.vote_totals[winner]
-    bp_winner_votes = bp_stratum.vote_totals[winner]
-
-    cvr_loser_votes = cvr_stratum.vote_totals[loser]
-    bp_loser_votes = bp_stratum.vote_totals[loser]
-
-    V = cvr_winner_votes - cvr_loser_votes + bp_winner_votes - bp_loser_votes
-
-    # The election is tied
-    if V == 0:
-        return 1.0
-
-    lambda_lower = (
-        np.amax(
-            [
-                cvr_winner_votes - cvr_loser_votes - cvr_stratum.num_ballots,
-                V - (bp_winner_votes - bp_loser_votes + bp_stratum.num_ballots),
-            ]
-        )
-        / V
-    )
-    lambda_upper = (
-        np.amax(
-            [
-                cvr_winner_votes - cvr_loser_votes + cvr_stratum.num_ballots,
-                V - (bp_winner_votes - bp_loser_votes - bp_stratum.num_ballots),
-            ]
-        )
-        / V
-    )
-
-    sample = bravo.compute_cumulative_sample(bp_stratum.sample)
-    bp_sample_winner_votes = sample[winner]
-    bp_sample_loser_votes = sample[loser]
-
-    reported_margin = contest.candidates[winner] - contest.candidates[loser]
-
-    Wn = bp_sample_winner_votes
-    Ln = bp_sample_loser_votes
-    Un = bp_stratum.sample_size - bp_sample_winner_votes - bp_sample_loser_votes
-    assert Wn >= 0, f"{Wn, Ln, Un}"
-    assert Ln >= 0, f"{Wn, Ln, Un}"
-    assert Un >= 0, f"{Wn, Ln, Un}"
-
-    T2 = (
-        lambda delta: 2
-        * bp_stratum.sample_size
-        * np.log(1 + reported_margin * delta / (2 * bp_stratum.num_ballots * GAMMA))
-    )
-    modulus = (
-        lambda delta: 2 * Wn * np.log(1 + reported_margin * delta)
-        + 2 * Ln * np.log(1 + reported_margin * delta)
-        + 2 * Un * np.log(1 + 2 * reported_margin * delta)
-        + T2(delta)
-    )
-
-    while True:
-        test_lambdas = np.arange(lambda_lower, lambda_upper + stepsize, stepsize)
-        if len(test_lambdas) < 5:
-            stepsize = (lambda_upper + 1 - lambda_lower) / 5
-            test_lambdas = np.arange(lambda_lower, lambda_upper + stepsize, stepsize)
-
-        fisher_pvalues = np.empty_like(test_lambdas)
-        for i, test_lambda in enumerate(test_lambdas):
-            try:
-                pvalue1 = np.min(
-                    [
-                        1,
-                        cvr_stratum.compute_pvalue(
-                            reported_margin, winner, loser, test_lambda
-                        ),
-                    ]
-                )
-            except ValueError:
```

Comment:
when would we see a ValueError and why return 0 in that case?

Reply:
If lambda is overly large, it can cause the p-value calculation to try to take a log of a negative value. In that case, we're saying that the sample is so good that the alternative (i.e. the reported margin) is impossible. I guess that means we should be setting the p-value to 1 instead. Good catch.

Reply:
Should that just be handled within the p-value computing function?

Reply:
Fair point. I can fix that.